### ISSUE TITLE
The aim of this repo

### ISSUE THREAD

**Issue by taseenb** [taseenb/three.js-extensions#1]
Hello, 

I created this repository to collect, share and possibly maintain the best open source extensions available for [three.js](https://github.com/mrdoob/three.js/). 

Three.js is currently a powerful, large library, but a lot of open source code has been (and will be) written to further extend its features. The list is vast: extra features, shaders, post-processing effects, tools that simplify some tasks, reduce repetitive work or help with debug, test performance, structure large or specific types of projects, etc.

No maintained collection of extensions exists yet and three.js does not have any guideline for creating plugins. All this useful code is often hard to find (which is the main reason I created this repo) and also inconsistent (which is a secondary problem, but maybe worth discussing?).

The idea is to make the best extensions available to everyone and easy to find, thanks to previous selection and categorization, with the help of developers and artists active in the community.

How would you like this collection to be? Do you think an official plugin format should be defined? How?

Any comments, ideas, help?

Cheers

**Comment by jonobr1** [comment]
I'm confused. Is this a replica of what `three.js/examples/js` is?

**Comment by taseenb** [comment]
It should not. But, at the moment, one of the places you can find "extensions" (or "plugins", I don't know which word should be used) is actually within examples.

**Comment by jonobr1** [comment]
Got it, thanks for the clarification. Sounds like a lot of work. However, I'd definitely use it if it exists and was better than `examples/js` + Googling :+1:

**Comment by mattdesl** [comment]
Just my 2c: I prefer having things separate, easy to install, and easy to depend on. 

It's a bit tricky with threejs due to its handling of versioning and global scope, but it's worth it compared to old school "copy paste" approaches. Examples:

https://github.com/hughsk/three-effectcomposer
https://github.com/mattdesl/three-orbit-controls
https://github.com/Jam3/three-bmfont-text

With dependencies you can start building higher level abstractions like:  
https://github.com/Jam3/three-orbit-viewer

Another thing worth mentioning: a lot of "ThreeJS extensions" don't really need to be specific to ThreeJS.. Developing them independently makes them more useful to a variety of purposes (in addition to threejs apps). 

Anyways, just my thoughts. 😀 I am definitely +1 on "collections" and more docs to help people discover the ecosystem outside of "threejs/src/examples".

**Comment by taseenb** [comment]
@mattdesl yes, i also like the idea of easy dependencies, which is actually the main goal of plugins i guess. And also agree that Three extensions don't need to be specific to Three.
I am wondering if you have any opinion on an "ideal" Three extension format: how should it be for the user and how should it connect to the `THREE` object.

The approach i'm seeing at the moment, actually simple and clean, is:

`
// require extension if necessary
var Extension = require( 'Extension' );

// instantiate it with options
var ext = new Extension( {
    option0: 0.3,
    option1: 'maybe',
    option2: [ 1, 2, 3 ]
} );

// then?
`

If you know people that may be interested in this discussion, don't hesitate to let them know.
